---
title: "Align and Interpret"
author: "Jake Sauter"
date: "2/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## **BWA + STAR**

Write a script that will: (3pt)
    run BWA on one of the samples from the Gierlinski dataset
    run STAR on the same sample
        Remember those three checks after read alignment:
        Is it a BAM file?
        Is it sorted?
        Is it indexed?

### **Retrieving the Gierlinkski Dataset**

The Gierlinkski Dataset can be downloaded via the script linked below.

[slurm_gierlinski_download_and_qc_script.sh](https://raw.githubusercontent.com/jakesauter/ANGSD_STAR_BWA/main/scripts/slurm_gierlinski_download_and_qc_script.sh)


### Running BWA on One Sample

We will choose `ERR458497.fastq.gz` (`WT_1`, `Lane 5`) as
our sample to run as input to BWA. 


```{bash, eval=FALSE}
#! /bin/bash -l

#SBATCH  --partition=angsd_class
#SBATCH  --nodes=1
#SBATCH  --ntasks=1
#SBATCH  --job-name=gierlinski_bwa
#SBATCH  --time=06:00:00    # HH/MM/SS
#SBATCH  --mem=16G 
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err

spack  load  bwa@0 .7.15% gcc@6 .3.0

bwa  index -p sacCer3_BWAindex/sacCer3  sacCer3.fa

bwa  mem  referenceGenomes/sacCer3_BWAindex/sacCer3 gierlinski/fastq/ERR458493.fastq.gz > gierlinski/alignments/ERR458493.bwa.sam
```


### **Running STAR on the Same Sample**

```{bash, eval=FALSE}
#! /bin/bash -l

#SBATCH  --partition=angsd_class
#SBATCH  --nodes=1
#SBATCH  --ntasks=1
#SBATCH  --job-name=gierlinski_star
#SBATCH  --time=06:00:00    # HH/MM/SS
#SBATCH  --mem=16G 
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err

spack  load  star@2 .7.0e

STAR --runMode  genomeGenerate \
     --runThreadN 1 \
     --genomeDir  sacCer3_STARindex \
     --genomeFastaFiles  sacCer3.fa \
     --sjdbGTFfile  sacCer3.sgd.gtf \
     --sjdbOverhang  99
  
STAR --runMode  alignReads \
  --runThreadN 1 \
  --genomeDir  referenceGenomes/sacCer3_STARindex \
  --readFilesIn  gierlinski/fastq/ERR458493.fastq.gz \
  --readFilesCommand  zcat \
  --outFileNamePrefix  gierlinski/alignments/ERR458493. \
  --outSAMtype  BAM  SortedByCoordinate
```














